libraries/cbor/src/writer.rs

```diff
@@ -289,6 +290,52 @@ mod test {
         assert_eq!(write_return(value_map), Some(expected_cbor));
     }
 
+    #[test]
+    fn test_write_map_sorted() {
+        let sorted_map = cbor_map! {
+            0 => "a",
+            1 => "b",
+            -1 => "c",
+            -2 => "d",
+            b"a" => "e",
+            b"b" => "f",
+            "" => "g",
+            "c" => "h",
+        };
+        let unsorted_map = cbor_map! {
+            1 => "b",
+            -2 => "d",
+            b"b" => "f",
+            "c" => "h",
+            "" => "g",
+            b"a" => "e",
+            -1 => "c",
+            0 => "a",
+        };
+        assert_eq!(write_return(sorted_map), write_return(unsorted_map));
+    }
+
+    #[test]
+    fn test_write_map_duplicates() {
+        let unique_map = cbor_map! {
+            0 => "a",
+            -1 => "c",
+            b"a" => "e",
+            "c" => "h",
+        };
+        let duplicate_map = cbor_map! {
```

Comment:
I think map duplicate keys are invalid CBOR. Otherwise it can't be canonical :)
This should be a rejected input.
Especially here, your test case has the same values for each key but appearing twice but if values were different, there would be no way figuring out which one is correct.

Reply:
Just making sure this is the failure behaviour we want, as it changes how we treat these cases. Generally, I am happy to introduce a failure here. Before, we simply overwrote and ignored duplicates.

This is the writer, so basically, `sort_by` and `dedup_by` are always no-ops. Sorting still makes sense to me, as the API is much nicer if we don't require sorted input (and should be fast for sorted input). Expecting dedup'ed input is reasonable, and has been an implicit property when we received `BTreeMap`. Silently ignoring duplicates best-effort means we catch bugs later, but helps ignore cases where we write the same key/value pair twice.

WDYT?

Reply:
I pushed a commit to change the behaviour, anticipating that this is what we will converge to. :)